paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkConnectorOptions.java

```diff
@@ -181,6 +181,12 @@ public class FlinkConnectorOptions {
                             "How many splits should assign to subtask per batch in StaticFileStoreSplitEnumerator "
                                     + "to avoid exceed `akka.framesize` limit.");
 
+    public static final ConfigOption<Boolean> LOOKUP_ASYNC =
```

Comment:
I feel a bit confused about this feature. From the FLIP and docs, we can see the join-hint will prefer async lookup join when we add hint `LOOKUP('table'='dim1', 'async'='true')`. 
However, if we take a look at the implementation:
[LookupJoinUtil#isAsyncLookup](https://github.com/apache/flink/blob/678370b18e1b6c4a23e5ce08f8efd05675a0cc17/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/utils/LookupJoinUtil.java#L412-L413), it will return async enabled via judging whether the created `lookupRuntimeProvider` is an instance of `AsyncLookupFunctionProvider`.
However we would only [return the sync lookup](https://github.com/apache/flink/blob/678370b18e1b6c4a23e5ce08f8efd05675a0cc17/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/utils/LookupJoinUtil.java#L614-L626) runtime provider without async option in connector DDL. 
In other words, if we did not give async related options in the connector DDL, the `LOOKUP('async'='true')` hint would not take effect.

Reply:
CC @lincoln-lil

Reply:
@Myasuka @JingsongLi  sorry for the late response. For flink planner, it does prefer the async lookup and will check what kind of lookup function does the backend source can provide 
```
// prefer (not require) by default
boolean preferAsync = preferAsync(lookupHint);
// check lookup function provider ...
return preferAsync ? asyncFound : !syncFound;
```
so, if the connector can have both sync and async lookup capabilities, it‘s better to implement both providers

Reply:
@lincoln-lil The `preferAsync` still need the `asyncFound` to return `true` for async-lookup. That is to say, we still need the instantiated provider is an `AsyncLookupFunctionProvider`. If so, we have to introduce anther connector-scope parameter of `async-lookup` to instantiate the provider. On the other hand, @JingsongLi thought we could leverage the lookup hint instead of introducing another parameter. That's the point to confuse me.

Reply:
@Myasuka I see your confusion. With the current table interfaces, the connector itself needs to provide a `LookupRuntimeProvider` that is both an `AsyncLookupRuntimeProvider` and a `LookupFunctionProvider` to achieve this goal, like this sample code:
```java
/** A full capability lookup table source. */
public static class MixedLookupTableSource implements LookupTableSource {

        @Override
        public LookupRuntimeProvider getLookupRuntimeProvider(LookupContext context) {
            AsyncLookupFunction asyncLookupFunction = new MyAsyncLookupFunction...
            LookupFunction lookupFunction = new MyLookupFunction...
            return MixedLookupFunctionProvider.of(asyncLookupFunction, lookupFunction);
        }

        interface MixedLookupFunctionProvider
                extends AsyncLookupFunctionProvider, LookupFunctionProvider {
            static MixedLookupFunctionProvider of(AsyncLookupFunction asyncLookupFunction, LookupFunction lookupFunction) {
                return new MixedLookupFunctionProvider() {
                    @Override
                    public AsyncLookupFunction createAsyncLookupFunction() {
                        return asyncLookupFunction;
                    }

                    @Override
                    public LookupFunction createLookupFunction() {
                        return lookupFunction;
                    }
                };
            }
        }
```
Yes, for now it looks like this part of the code might be easier for the connector developer if the planner can figure out how to handle it, such as adding a boolean preferAsync() to the `LookupContext`, which would also save the overhead of creating a new function, I'll track this in flink.